### DISCUSSION TITLE
Dynamic States for a Wizard

### DISCUSSION THREAD

**Discussion by tobiastimm** [statelyai/xstate#1654] [Q&A] [answered]
Hey all 👋 
I wanted to implement a wizard with `xstate` that can have a dynamic amount of steps. Any ideas on how to implement that?

Depending on the user choice, I can have one or more additional steps added to the wizard. I have this information before I initialize the wizard.

At the moment, I have two ideas to approach this problem:

1. Explicitly write down all possible steps and skip some based on certain conditions
For this one, I would explicitly write every possible step as a state in the machine. Based on certain conditions I would try to skip some steps.

2. Create a step machine and call that x amount of times
The needed work for a step is repetitive; that's why I thought about a step machine
```
const stepMachine = Machine({
  id: 'step',
  initial: 'idle',
  context: {
    stepName: '',
  },
  states: {
    active: {
      states: {
        invalid: {
          on: {
            VALIDATE: {
              target: 'valid',
            },
          },
        },
        valid: {
          on: {
            NEXT: '#completed',
          },
        },
      },
      on: {
        BACK: '' // send to Parent and go back to previous step,
      },
    },
    completed: {
      type: 'final',
    },
  },
});
```
Within the parent, I would call this state x amount of times till every step is valid and complete; afterward, I would enhance to the overall completed state

Thank you in advance for any help 😄

**Comment by davidkpiano** [comment, marked as answer]
Instead of "hardcoding" each of the steps, consider putting them in `context`: https://codesandbox.io/s/xstate-react-back-example-4q2vh

**Reply by d-sanderson** [reply]
How would you handle adding guarded transitions that are unique to each step?

**Comment by tobiastimm** [comment]
@davidkpiano Thank you very much for that example! I will try it out 👍
